### PR TITLE
[CHIA-435] Add `chia wallet signer` commands

### PR DESCRIPTION
**Pull request by Quexington**
This PR caps off a thread of work that supports external signers for transactions generated by the wallet.  With these commands, it should be possible to run two wallets with different security qualities and transport transactions back and forth between them for the colder one to sign.  It currently uses binary files/QR codes to accomplish this although other methods of transport are certainly on the table.

### CODE REVIEW

File: chia/cmds/chia.py [hunk 1]
```diff
@@ -134,6 +134,8 @@ def run_daemon_cmd(ctx: click.Context, wait_for_unlock: bool) -> None:
 
 
 def main() -> None:
+    import chia.cmds.signer  # noqa
```
Comment (altendky) [inline chia/cmds/chia.py:137]: What side effect is this being imported for?

**Review by altendky — COMMENTED** [review]
a spattering of thoughts

File: chia/cmds/cmd_classes.py [hunk 2]
```diff
@@ -0,0 +1,327 @@
...
+        metadata=dict(
+            is_command_option=True,
+            param_decls=tuple(param_decls),
+            **kwargs,
+        ),
+        default=kwargs["default"] if "default" in kwargs else default_default,
+    )
+
+
+class HexString(click.ParamType):
```
[the remark this answers is not in the document]
Comment (Quexington) [reply chia/cmds/cmd_classes.py:66]: Ah yeah I was not aware of that my bad.  This, however is just for non-32 byte length hex strings.

File: chia/cmds/cmd_classes.py [hunk 3]
```diff
@@ -0,0 +1,327 @@
...
+        default_default = None  # pragma: no cover
+    else:
+        default_default = MISSING
+
+    return field(  # pylint: disable=invalid-field-call
+        metadata=dict(
+            is_command_option=True,
+            param_decls=tuple(param_decls),
+            **kwargs,
+        ),
```
[the remark this answers is not in the document]
Comment (Quexington) [reply chia/cmds/cmd_classes.py:61]: This function that returns a `field` is designed to basically forward all arguments to the `click.option` decorator.  That way devs making commands can still refer to click documentation.  We just for the sake of our own processing need to store this as a field on the dataclass so we have somewhere to read it from when we are dynamically applying the decorators.

Using a subkey is a good suggestion though so if you wanted to extend this framework or force some other behavior, you have a hook.

File: chia/cmds/cmd_classes.py [hunk 4]
```diff
@@ -0,0 +1,327 @@
...
+    return _chia_command
+
+
+class _CommandHelper:
+    pass
+
+
+@dataclass_transform()
+def command_helper(cls: Type[Any]) -> Type[Any]:
+    if sys.version_info < (3, 10):  # stuff below 3.10 doesn't support kw_only
```
[the remark this answers is not in the document]
Comment (Quexington) [reply chia/cmds/cmd_classes.py:273]: It shouldn't no, obviously if you don't specify `kw_only` you can still _in practice_ only use kwargs and that practice will be enforced by our CI which runs 3.10+.  However, when running on lower versions (that we support) the `@dataclass` decorator doesn't know about the `kw_only` argument and would error if we don't omit it.

File: chia/cmds/cmd_classes.py [hunk 5]
```diff
@@ -0,0 +1,327 @@
...
+        return dataclass(frozen=True)(
+            type(cls.__name__, (dataclass(frozen=True)(cls), _CommandHelper), {})
+        )  # pragma: no cover
+    else:
+        return dataclass(frozen=True, kw_only=True)(
+            type(cls.__name__, (dataclass(frozen=True, kw_only=True)(cls), _CommandHelper), {})
+        )
```
[the remark this answers is not in the document]
Comment (Quexington) [reply chia/cmds/cmd_classes.py:280]: Yes this is a good idea, I got a little bit down of a rabbit hole here I think.

[the remark this answers is not in the document]
Comment (Quexington) [reply chia/cmds/cmd_classes.py:280]: Yes this is a good idea, I got a little bit down of a rabbit hole here I think.

File: chia/cmds/cmd_classes.py [hunk 6]
```diff
@@ -0,0 +1,327 @@
...
+    else:
+        default_default = MISSING
+
+    return field(  # pylint: disable=invalid-field-call
+        metadata=dict(
+            is_command_option=True,
+            param_decls=tuple(param_decls),
+            **kwargs,
+        ),
+        default=kwargs["default"] if "default" in kwargs else default_default,
```
[the remark this answers is not in the document]
Comment (Quexington) [reply chia/cmds/cmd_classes.py:62]: The special casing is actually due to the fact that dataclasses will raise if you follow a default argument with a non-default argument.  This means you have to specify all of your options with defaults at the end which is quite weird and magic.  So we get around this by always specifying a default (including `MISSING`) so you can put your options in whatever order you want.

[the remark this answers is not in the document]
Comment (Quexington) [reply chia/cmds/cmd_classes.py:62]: I'll apply the first suggestion though

File: chia/cmds/cmd_classes.py (hunk 3 above)
[the remark this answers is not in the document]
Comment (Quexington) [reply chia/cmds/cmd_classes.py]: Subkey utilized

File: chia/cmds/cmd_classes.py (hunk 5 above)
[the remark this answers is not in the document]
Comment (Quexington) [reply chia/cmds/cmd_classes.py]: Simplified

File: chia/cmds/cmd_classes.py (hunk 6 above)
[the remark this answers is not in the document]
Comment (altendky) [reply chia/cmds/cmd_classes.py]: Interesting work around, yeah.  Though it's less weird and magic if you consider that the order of attributes is the order of init arguments which matters when you are passing by position.  Which yes, you aim to ban via `kw_only` when possible below.

[the remark this answers is not in the document]
Comment (altendky) [reply chia/cmds/cmd_classes.py]: Maybe I don't quite understand, but I'm not sure this is actually doing the work around.  Below still results in a `TypeError`.  `field()` doesn't actually know if you passed `default=MISSING` or not and it just got the parameter default.  I think it _could_ use `inspect` and figure it out, but i don't think it bothers.

```python
from dataclasses import MISSING, dataclass, field


try:
    @dataclass
    class C:
        a: int = 0
        b: str
except TypeError:
    pass
else:
    assert False, "dataclass should have failed"

@dataclass
class C:
    a: int = 0
    b: str = field(default=MISSING)
```

[the remark this answers is not in the document]
Comment (Quexington) [reply chia/cmds/cmd_classes.py]: Yeah sorry, I misremembered a little bit in my explanation.  It's not about always specifying a default but rather always specifying a default _on the <3.10 versions_.  In your example, if you change that decorator to use `kw_only=True` you don't get an error.

**Review by geoffwalmsley — APPROVED** [review]

**Review by wjblanke — APPROVED** [review]
aok

**Pull request merged**